### PR TITLE
Refactor client/index.js for better testability

### PR DESCRIPTION
**Pull request by davelosert**
PR for #594 .

As stated I extracted a factory functions and added several more tests for the paths given to the `bundleClient` and `fs` functions.

I also added a first test for the paths that get built within the `client/index.js`. I realized quickly that thre is made use of `require.resolve` - which is not mockable and therefore the built path is not testable right now.

`require.resolve` is also used within the `server/public.js` file. I think it could be a good idea to take all those `require.resolve` calls into an own separate module like `modulePathResolver` that could look like this:

```
var path = require('path')
var modulePathResolver = {
   getClientPath() : function () {
      return require.resolve('@hoodie/client/package.json')
  }
  getPublicPath : function () {
    return path.join(require.resolve('../../package.json'), '..', 'public')
  }
  getAccountPublicPath : function () {
    return path.join(require.resolve('@hoodie/account/package.json'), '..', 'public')
  }
  // ... and so on for all modulePaths
}
```

@gr2m What do you think about that? Should I raise another issue for it?

### CODE REVIEW

**Comment by gr2m** [comment]
> I think it could be a good idea to take all those require.resolve calls into an own separate module like modulePathResolver

Good idea 👍

File: server/plugins/client/bundle-handler-factory.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
+module.exports.createHandler = createHandler
```
Comment (gr2m) [inline server/plugins/client/bundle-handler-factory.js]: why not simply `module.exports = createHandler`? That’s what we usually do: 1 file - 1 method :)

File: server/plugins/client/bundle-handler-factory.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+          resolve(bundleBuffer)
+        })
+      })
+    }
+
+    bundlePromise.then(function (buffer) {
+      reply(buffer).bytes(buffer.length).type('application/javascript')
+    }).catch(reply)
+  }
+}
```
Comment (gr2m) [inline server/plugins/client/bundle-handler-factory.js:31]: We usually don’t do factories, because they are harder to test. We usually prefer pre-binded methods (but that doesn’t mean that it’s always the right way, this is just a suggestion, up for discussion :)

``` js
module.exports = createHandler
var bundlePromise

function createHandler (hoodieClientPath, bundleTargetPath, bundleConfig, request, reply) {
  if (!bundlePromise) {
    bundlePromise = new Promise(function (resolve, reject) {
      log.silly('client', 'bundling')
      bundleClient(hoodieClientPath, bundleTargetPath, bundleConfig, function (error, bundleBuffer, hasUpdate) {
        if (error) {
          return reject(error)
        }

        writeClientBundle(hasUpdate, bundleConfig.inMemory, bundleTargetPath, bundleBuffer)

        resolve(bundleBuffer)
      })
    })
  }

  bundlePromise.then(function (buffer) {
    reply(buffer).bytes(buffer.length).type('application/javascript')
  }).catch(reply)
}
```

And then in `server/plugins/client/index.js` we could do 

``` js
  server.route([{
    method: 'GET',
    path: '/hoodie/client.js',
    handler: createBundleHandler.bind(null, hoodieClientPath, bundleTargetPath, options.config)
  }])
```

File: server/plugins/client/bundle-handler-factory.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  log.silly('client', 'writing bundle to ' + bundleTargetPath)
+  fs.writeFile(bundleTargetPath, bundleBuffer, function (error) {
+    if (error) {
+      log.warn('client', 'could not write to ' + bundleTargetPath + '. Bundle cannot be cached and will be re-generated on server restart')
+      return
+    }
+
+    log.info('client', 'bundle written to ' + bundleTargetPath)
+  })
+}
```
Comment (gr2m) [inline server/plugins/client/bundle-handler-factory.js]: we could also move this into its own file or exporting at `module.exports.internals.writeClientBundle` for unit testing?

**Review by gr2m — CHANGES_REQUESTED** [review]
This is a great pull request, thanks so much for your great work! I’ve some feedback on how we could maybe simplify implementation and testing, up for discussion :)

File: server/plugins/client/bundle-handler-factory.js (hunk 2 above)
Comment (davelosert) [reply server/plugins/client/bundle-handler-factory.js:31]: I usually prefer factories for the reason as the word 'factory' itself already makes it pretty clear what they are doing. :)

What weighs even stronger is consistency with the rest of the code. In this particular case we have to consider one thing though: The `bundlePromise` - variable in your example is kind of "global" within the module. I don't assume it is a use case right now, but if you would `require` the createHandler twice from different clients, the second handler would already find the `bundlePromise` not being undefined and therefore never call `bundleClient` (with a maybe different configuration than the first). 

You still think we should do it like this? Or maybe do it with `bind` but still wrap the "createHandler" into a function for a private scope?

File: server/plugins/client/bundle-handler-factory.js (hunk 3 above)
Comment (davelosert) [reply server/plugins/client/bundle-handler-factory.js]: Haha, I was actually thinking about doing this but then thought it is not that complex to test right now.
But of course, it would make things even easier! Shall I do it within this PR or open another issue after we are done here?

File: server/plugins/client/bundle-handler-factory.js (hunk 1 above)
Comment (davelosert) [reply server/plugins/client/bundle-handler-factory.js]: Ah - i actually oriented myself in how it was done within the `index.js`:

```
module.exports.register = register
module.exports.register.attributes = {
  name: 'hoodie-local-client',
  dependencies: 'inert'
}
```

But maybe it's different there because it's the index and you add some more attributes to it?
Easy change I can make, though :)

File: server/plugins/client/bundle-handler-factory.js (hunk 2 above)
Comment (gr2m) [reply server/plugins/client/bundle-handler-factory.js:31]: you are right, and I think we need to bundle more than just `hoodie.js` soon. We could probably combine the two, have a factory that does nothing but defining the bundle variable and then pass it into a pre-bound function like my suggestion above? So the factory method would be only two lines, and we can test the pre-bound function quite easily if we want to

File: server/plugins/client/bundle-handler-factory.js (hunk 1 above)
Comment (gr2m) [reply server/plugins/client/bundle-handler-factory.js]: `index.js` is because it’s a hapi plugin, which has certain requirement: http://hapijs.com/tutorials/plugins

File: server/plugins/client/bundle-handler-factory.js (hunk 3 above)
Comment (gr2m) [reply server/plugins/client/bundle-handler-factory.js]: I think you can make this part of this PR

File: server/plugins/client/bundle-handler-factory.js (hunk 2 above)
Comment (davelosert) [reply server/plugins/client/bundle-handler-factory.js:31]: You mean something like this?

```
function createHandler(hoodieClientPath, bundleTargetPath, bundleConfig){
   var bundlePromise
   return bundleHandler.bind(null, hoodieClientPath, bundleTargetPath, bundleConfig, bundlePromise)
}

function createHandler (hoodieClientPath, bundleTargetPath, bundleConfig, bundlePromise, request, reply) {
  if (!bundlePromise) {
      ...
  }
```

In that case the `bundlePromise` within the factory-function won't every be changed as it isn't passed by reference, is it? Or did you mean it differently?

File: test/unit/client/client-test.js [hunk 4]
```diff
@@ -4,32 +4,20 @@ var test = require('tap').test
 
 require('npmlog').level = 'error'
 
-test('client', function (group) {
+test('index', function (group) {
```
Comment (gr2m) [inline test/unit/client/client-test.js]: why did you change it from client to index?

**Review by gr2m — CHANGES_REQUESTED** [review]

**Comment by davelosert** [comment]
@gr2m I squashed the commits now. As soon as your saucelabs tests are working again I'd say this is ready to be merged! :)

**Comment by gr2m** [comment]
Thanks 👍 I still didn’t figure out the selenium / saucelabs problems but I have it on my list

**Comment by gr2m** [comment]
just a short update, I did not forget about the PR, we still did not resolve the selenium tests, sorry it takes so long to finish up this PR

**Comment by davelosert** [comment]
No problem. Just merge it as soon as it's ready ;)

**Review by gr2m — APPROVED** [review]
all right, I could finally fix the tests of the tracker app and it all looks good now. I rebased your PR on the latest `master` branch and will merge it once CI is green. Thanks so much for your great work 🙌

**Comment by davelosert** [comment]
:hand: great!

**Comment by gr2m** [comment]
✋ 💯
